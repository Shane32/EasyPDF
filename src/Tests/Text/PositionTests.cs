using System.Drawing;
using Font = Shane32.EasyPDF.Font;

namespace Tests.Text;

public class PositionTests
{
    private readonly PDFWriter _writer = new PDFWriter();

    public PositionTests()
    {
        _writer.ScaleMode = ScaleModes.Inches;
        _writer.NewPage(System.Drawing.Printing.PaperKind.Letter, false, 1f, 1f);
        _writer.PrepForTests();
    }

    [Theory]
    [InlineData(1f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0f, 1f, 0f, 0.875f)]
    [InlineData(1.2f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.18f, 0.03f, 0f, 1f, 0f, 0.875f)]
    [InlineData(1f, 6f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0.083f, 1f, 0f, 0.875f)]
    [InlineData(1f, 0f, 1.2f, 0.1324f, 0.1366f, 0.043f, 0.18f, 0f, 0f, 1f, 0f, 0.875f)]
    [InlineData(1.2f, 6f, 1.2f, 0.1324f, 0.1366f, 0.043f, 0.216f, 0.036f, 0.083f, 1f, 0f, 0.875f)]
    [InlineData(1f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0f, 1.2f, 0f, 1.05f)]
    [InlineData(1f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0f, 1f, 1f, 1.027f)]
    [InlineData(1f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0f, 1.2f, 1f, 1.233f)]
    public void TextSizes(float lineSpacing, float paragraphSpacing, float stretchY, float capHeight, float ascent, float descent, float height, float leading, float paragraphLeading, float stretchX, float characterSpacing, float width)
    {
        _writer.Font = new Font(StandardFonts.Times, 12f) {
            LineSpacing = lineSpacing,
            ParagraphSpacing = paragraphSpacing,
            StretchY = stretchY,
            CharacterSpacing = characterSpacing,
            StretchX = stretchX,
        };
        _writer.TextCapHeight().ShouldBe(capHeight, 0.001f);
        _writer.TextAscent().ShouldBe(ascent, 0.001f);
        _writer.TextDescent().ShouldBe(descent, 0.001f);
        _writer.TextHeight().ShouldBe(height, 0.001f);
        _writer.TextLeading().ShouldBe(leading, 0.001f);
        _writer.TextHeight().ShouldBe(_writer.TextAscent() + _writer.TextDescent() + _writer.TextLeading(), 0.001f);
        _writer.TextParagraphLeading().ShouldBe(paragraphLeading, 0.001f);
        _writer.TextWidth("Hello World!").ShouldBe(width, 0.001f);
    }
    
    [Fact]
    public void Standard()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f);
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void StretchedY()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 5f) { StretchY = 2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void Spacing()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f) { LineSpacing = 1.2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void StretchedY_Spacing()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 5f) { StretchY = 2f, LineSpacing = 1.2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void StretchedX()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f) { StretchX = 2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void CharacterSpacing()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f) { CharacterSpacing = 2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void MultiLine()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f);
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.5f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment + "\nLine 2");
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

}
