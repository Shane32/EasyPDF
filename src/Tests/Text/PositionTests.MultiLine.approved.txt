%PDF-1.4
%\xE2\xE3\xCF\xD3
2 0 obj
<</Length 1024/Filter/FlateDecode>>stream
x\x9C\x95W\xCBn\x1C9\x0C\xBC\xF7W\xF4\xD1>\xB8#\xEA\xD5Rn\x9B\xC5&@\x90K\x92\xB9\x05>\x04\xCE\xC4\xF1\xC2c\xE7a \xBF\x1F\xB5H\x8A\xEC\xC0\x86%\xD8\xC6pZd\x89\xAA\x9AQ\x99?\xA6W\x87\xC9\xC5yq>|\x99\xFE;L\xEF'\x98M\xF9\xB9\x80y\xB5\xE5\xD7\xCCW\xA7\xC9\xCCo\xCB\xDF\xFF\x93Y`\xFE=}\xBA,\xEB_&S\xF3>\xBC\xA1\xE0\xE7\xF5\xE3i\xD6\x9A%\x94\xE8D\x91]|\x9E-\xA4\xED\xC5o\xEF!\xD6\xD7\xAB\xC9\x82[\x02\xD0C\x90T\xA8\xF5\xDB\xFA\x16]\xD0\xD3\x9A{\xC1\x08\x17\x0CQq\xF1q\xDD\x8E\xD2\xA9\x89\xAB\xE9\xE3\xC4\x80\xA7\xD6\xDAm}J \xA7\x89;\xDA\x1E\xFF(\x07:\\m,\xBDx3\x947_\x85\xA0-q] \xCD\x87\x93&\xC1lT~:\xFB\xE7\xF6\xE6\xFA\xEEt\xBC{xy\x1E\xCC\xD9\xFC\xEE\xF8\xF5\xE1p\x0E\xD6\x9C\xDD\x7F?\xBF<\xBCE\xB2\x9F\xC5\x87\xB8x\xF7\xF8\x06\xEFn\xEE\x8E\xB3\x15,<NQ\x93O\xE6\x92\xE6\xDA0\xD9f\xC7\xB6its: 
\x13\xEEj\x1E\xA5; \x94\x1A4\xC2\xF11n\x8A\xE9\xDC\x8B0\xAE\x1B\x8B\x8D\xF3ZI\x9C\x9B^\xD2]^B\xEEf\xFD\xDF\xF2\xEE\xF8s\x80s\x9F\x96\xE4G8/\xDF\x14>\xDA\xEA\x15\xE7k$\xB6j\xD08\xA7\xC7\xA0\xD2\x01Q\x98\xF3\x98\x15\xE7q%\x94\x1A4\xCE\xF1q\xDD\x94\xD2\xB9\x17\xE1\\7f\x1B\xE7\xB5\x129\xAF\xAD\xF4}\xD0m7\xE1\xAF>\xFF:\xDE\x16\xA2\x06(O\xB0\xD80B9\x98\xD4\x8E\x06`\x14\xE9\x00\x96\xF8\xC2\xA8\xD1\xCE\x0B\xA0K\x80\xB0\x98y0AQ\x0F\xC61V\x8D\x1A\xF9\xB4@\xAD\x04\xA1\x1F\xB1>N\x82|R\x1D\xB3\x02X\x8E\x12`[]\x1Al;\xE5'n\x83\xC7t\xB8\x7Fx\xB8?\x8D\\6\x10\x9E\xBC\xCE\x9E\x90\xC1{9\xA0\x8FZ\x86\xF2-"\xEAj$2\xD0\x02\xE8\x12 \xAC&\x83\x07-\xC3\xF6\x85G\xAC\x1A\x89\x0C\xB8@\xDB\x83\x92\xA1b5\x19v]z\x91\xA1\x96\x93\x0C\xB5\xADge\x80\x98\x16kg\x08\xD0{\xF1\xE3\xF53z\xF5\x1B\xBB\xB8Tv3\x83\xD7?$q6H\xDAl!\xB1Yb$\x82$\xB1\xDCV\x02\x84\xD5\x04YW-\xC8\x1A\x18\xABF"\x08.\xE0\xF6T\xD2\xFAR\x82\xEC\xBA\x14\x07\xC6r\x12$u\x990\xC4ru\xB8\x92\xEDz=\x01\x05\x19s\x05\x92#\xDBAg\xD8\x0E\xD2\xFE\xF3\x01\xED\xC7\xB2\x8EQ\x93\x83\x17@\x97\xB0\xED\xB2\x1C\x9C\xCB/\xBC\x8EQ\x93\x83\x16\x08q\x07\xAF\xADy\xDF\xA5\x983\xAE\xA2\x1C\xB8\xFA\xBC\x1Ca]L\xAC(\xFDZ\x8C\x1A\x06\xAAam\x184\x1B\xC4\x0Fm\xD0Nm\x03[5F\xA2F\x10\xB3n%@XM\xAF\xFD\xDAz6l\x8CD/\x96\xCD%\xAD/\xA5\xC6\xAEK\xB1m,'5B\x97qo\xF7\x87\xCBu\xB3N\xDF A\x86\x9C\x83\xE4\x08y\xD0<l\x12w\xB4Y{\xB8\xCD\xEC\xE1\x18\x89\x1CY<\xBC\x95\x00a59\x92\xF6p\x9B\xD8\xC31\x129\x92x8\x97\xB4\xBE\x94\x1C\xBA\xCB$\x1E\x8E\xE5$G\xEE\xF2p\xB0e\\\xD8\x0E\xF1\xB4\xCD\xFE%\xC7\x87\x9B\xEBo\xA3c\x03\xA4\xB4\xAC\xE5\x9Fd\xE3GG\x07+6\xE9\xAC6sg\xD9\xCC1jz\xF0\x02\xE8\x12 \xAC6AX\xD8\x8D\x10l\xE6\x18\xA9!B\xCC\x9CKZ_j\x8E\xD8u\xE9\xD5$!f\x8Em=\xAFG\xE9\x08\\\xC9^{\xBD\xA3\xEA1d\x1D\xAC\x86\x8B\x83\xD6\xE1\xA2Q\x03\x93vr\x17\xD9\xC91\x125\xA28y+\x01\xC2jj\x04\xED\xE4.\xB0\x93c$j\x04qr.i}\xE9\xA9Nw)N\x8E\xE5\xA4F\xECs\xF2\xF2\x81]sE\xE9\x96b\xD09X\x8C20\x8D9\x87\xCBjz\xCD\xFB\xB9Z\x06\xEB\xBF&k5Zg=[g5\\\xE7\xDDt\x9D\xDBx\x9D\xF7\xF3uV\x03v\xD6\x13v\xDE\x8D\xD8\xBB.\xD5\x90\x9D\xF5\x94\xDD\xE7\xE3\x05\xD0\xC7\xBAY\xA7s\xA0\x1E#\xC6Aj\xF82\x0C\xF5\x1A\xC7\x1F\xA5\xEC\xD6\x01
endstream
endobj
4 0 obj
<</Type/Page/MediaBox[0 0 612 792]/Resources<</ProcSet [/PDF /Text /ImageB /ImageC /ImageI]/Font<</F1 1 0 R>>>>/Contents 2 0 R/Parent 3 0 R>>
endobj
1 0 obj
<</Type/Font/Subtype/Type1/BaseFont/Helvetica/Encoding/WinAnsiEncoding>>
endobj
3 0 obj
<</Type/Pages/Count 1/Kids[4 0 R]/ITXT(3.3.2.0)>>
endobj
5 0 obj
<</Type/Catalog/Pages 3 0 R>>
endobj
6 0 obj
<<>>
endobj
xref
0 7
0000000000 65535 f 
0000001264 00000 n 
0000000015 00000 n 
0000001352 00000 n 
0000001107 00000 n 
0000001418 00000 n 
0000001463 00000 n 
trailer
<</Size 7/Root 5 0 R/Info 6 0 